gen.go

```diff
@@ -0,0 +1,215 @@
+	switch t := schema.(type) {
+	case *RecordSchema:
+		typeName := strcase.ToCamel(t.Name())
+		fields := make([]*ast.Field, len(t.fields))
+		for i, f := range t.fields {
+			fSchema := f.Type()
+			fieldName := strcase.ToCamel(f.Name())
+			typ := resolveType(fSchema, f.Prop("logicalType"), acc)
+			tag := f.Name()
+			fields[i] = newField(fieldName, typ, tag)
+		}
+		acc.Decls = append(acc.Decls, newType(typeName, fields))
+		return typeName
+	default:
+		return resolveType(schema, nil, acc)
+	}
+}
+
+func resolveType(fieldSchema Schema, logicalType interface{}, acc *ast.File) string {
+	var typ string
+	switch s := fieldSchema.(type) {
+	case *RefSchema:
+		typ = strcase.ToCamel(s.actual.Name())
+	case *RecordSchema:
+		typ = generateFrom(s, acc)
+	case *PrimitiveSchema:
+		typ = resolvePrimitiveLogicalType(logicalType, typ, s)
+		if strings.Contains(typ, "time") {
+			addImport(acc, "time")
+		}
+		if strings.Contains(typ, "big") {
+			addImport(acc, "math/big")
+		}
+	case *ArraySchema:
+		typ = fmt.Sprintf("[]%s", generateFrom(s.Items(), acc))
+	case *EnumSchema:
+		typ = "string"
+	case *FixedSchema:
+		typ = fmt.Sprintf("[%d]byte", +s.Size())
+	case *MapSchema:
+		typ = "map[string]" + resolveType(s.Values(), nil, acc)
+	case *UnionSchema:
+		typ = resolveUnionTypes(s, acc)
+	}
+	return typ
+}
+
+func resolveUnionTypes(unionSchema *UnionSchema, acc *ast.File) string {
+	nullIsAllowed := false
+	typesInUnion := make([]string, 0)
+	for _, elementSchema := range unionSchema.Types() {
+		if _, ok := elementSchema.(*NullSchema); ok {
+			nullIsAllowed = true
+		} else {
+			typesInUnion = append(typesInUnion, generateFrom(elementSchema, acc))
+		}
+	}
+	if nullIsAllowed && len(typesInUnion) == 1 {
+		typ := typesInUnion[0]
+		if strings.HasPrefix(typ, "[]") {
+			return typ
+		}
+		return "*" + typ
+	}
+	return "interface{}"
+}
+
+func resolvePrimitiveLogicalType(logicalType interface{}, typ string, s Schema) string {
+	switch logicalType {
+	case "", nil:
+		typ = primitiveMappings[s.Type()]
+	case "date", "timestamp-millis", "timestamp-micros":
+		typ = "time.Time"
+	case "time-millis", "time-micros":
+		typ = "time.Duration"
+	case "decimal":
+		typ = "*big.Rat"
+	}
+	return typ
+}
+
+func newType(name string, fields []*ast.Field) *ast.GenDecl {
+	return &ast.GenDecl{
+		Tok: token.TYPE,
+		Specs: []ast.Spec{
+			&ast.TypeSpec{
+				Name: &ast.Ident{Name: name},
+				Type: &ast.StructType{
+					Fields: &ast.FieldList{
+						List: fields,
+					},
+				},
+			},
+		},
+	}
+}
+
+func newField(name string, typ string, tag string) *ast.Field {
+	return &ast.Field{
+		Names: []*ast.Ident{{Name: name}},
+		Type: &ast.Ident{
+			Name: typ,
+		},
+		Tag: &ast.BasicLit{
+			Value: "`avro:\"" + tag + "\"`",
+			Kind:  token.STRING,
+		},
+	}
+}
+
+func addImport(acc *ast.File, statement string) {
+	astutil.AddImport(token.NewFileSet(), acc, statement)
+}
+
+func writeFile(w io.Writer, f *ast.File) error {
```

Comment:
I wonder if a Go template would not be simpler here? Additionally the general "this code is generated" comment should probably be applied.

Just off the  top of my head, were a template used, a struct with the data could be made, including a Tag list, which could make that generation simpler. Just a thought.

Reply:
This felt simpler to me because doing for loops and ifs in Go code is (to me) simpler than doing them in Go templates. But I agree the calls to `Write` can hurt visibility and a template might be easier for anyone not familiar with `*ast.File`, just as I wasn't.

I'll make a commit with another version and we can look at both and re-evaluate which is better.

Reply:
[0956999](https://github.com/hamba/avro/pull/159/commits/0956999b072a9d7ce0185f62741386d4621d6afe) added this one, but then at this point the `ast.File` was just an indirection layer that didn't need to be there.

Originally I used it to explore it an because it allows more features for the future, but for now text based generation works fine, so I added [019b98b](https://github.com/hamba/avro/pull/159/commits/019b98b63dffb41c5cee58c51be86d0bb91e3277) to drop the references to `ast.File`.